### PR TITLE
Optimize event loop

### PR DESCRIPTION
**Pull request by ma-anwar**
## Describe your changes
The event loop previously would consume a large amount of CPU because it would always check if sockets are ready to write. A healthy socket is always ready for write, so this condition would always be true but we would have nothing to send.
Now, we toggle the selector to listen for write events once we put a message in the `out_buffer`. Then once the server sends the message, it toggles listening for write readiness off. 

Additionally, I fixed an age old bug that was undiscovered. I had previously done an incomplete migration to a generator method when dispatching messages to the parser that was faultily modifying the `out_buffer`, somehow everything kept working even with this bug.

Finally, I updated documentation.

## Testing
There are no regressions in integration tests and all functionality is still working with clients. 

## Checklist
- [x] I tested my code
- [x] I reviewed the diff
- [x] My code passed all pre commit hooks

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -7,7 +7,6 @@ readme = "README.md"
 
 [tool.poetry.dependencies]
 python = "^3.10"
-irc = "^20.1.0"
```
Comment (serhatgktp) [inline pyproject.toml:10]: Wasn't this a dependency? What did we use this for?

File: src/daemon/client.py [hunk 2]
```diff
@@ -544,3 +544,5 @@ def send_message(
             message_as_bytes += constants.IRC_TERMINATION_DELIMITER
 
         self._key.data.out_buffer += message_as_bytes
+        # Set event loop selector to listen for write readiness on socket
+        self._key.data.toggle_writable(self._key.data, True)
```
Comment (serhatgktp) [inline src/daemon/client.py:548]: I understand that we expect the client to write to socket after we send a message to the client.

When we broadcast a message to multiple users on a channel, we are calling each of their send_message functions, thus modifying their respective selectors to EVENT_WRITE.

Each of those selectors will only be toggled to EVENT_READ once we receive a message from the respective clients.

Is my interpretation correct?

**Review by serhatgktp — COMMENTED** [review]
This is a really cool fix!
I've left comments asking whether I'm interpreting the flow correctly. Please let me know!

File: pyproject.toml (hunk 1 above)
Comment (ma-anwar) [reply pyproject.toml:10]: For irc client that we removed

File: src/daemon/client.py (hunk 2 above)
Comment (ma-anwar) [reply src/daemon/client.py:548]: Good question, we will toggle the selector to listen for EVENT_WRITE on each socket yet. They will be toggled back to EVENT_READ once we are done writing the message onto the socket

**Pull request merged**